### PR TITLE
[GN] Flashing for nrf5 examples

### PR DESCRIPTION
**Pull request by kpschoedel**
Make-based builds for nrf5 lock-app and lighting-app had a `make flash` command to flash the device. GN builds have not had this functionality.

With this change, GN builds produce a Python script with the executable name followed by `.flash.py` which when run flashes an attached device in the same way that `make flash` does for Make builds. It is intended that this will evolve into a consistent method for all relevant build artifacts across all platforms, for developers and automated testing. The GN `flashable_executable()` template is platform-independent and extended with the platform specifics by `nrf5_executable()`.

The generated `.flash.py` script is a minimal wrapper that passes file names to a more generic nrf5 flashing utility script. Similarly, changes to the application BUILD.gn are minimal.

issue #1520 GN build parity - flashing

### CODE REVIEW

File: examples/lock-app/nrf5/make_flasher.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+      subprocess.check_call(nrfjprog + [
+        "--program",
+        os.path.join(os.path.dirname(sys.argv[0]), "{image}"),
+        "--sectorerase"])
+
+      subprocess.check_call(nrfjprog + ["--reset"])
+    """
+
+    content = ('#!/usr/bin/env python\n'
+               + '# Generated at {:%Y-%m-%d %H:%M:%S}'.format(
```
Comment (mspang) [inline examples/lock-app/nrf5/make_flasher.py:37]: Why add this nondeterminism?

File: examples/lock-app/nrf5/make_flasher.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+import sys
+import textwrap
+
+def main(argv):
+    parser = argparse.ArgumentParser(description='Flashing script generator.')
+
+    parser.add_argument("--output", metavar='FILENAME', required=True,
+                        help='flashing script name')
+    parser.add_argument("--image", metavar='FILENAME', required=True,
+                        help='image to be flashed')
```
Comment (mspang) [inline examples/lock-app/nrf5/make_flasher.py:16]: Should add a --flash-softdevice option probably!

And maybe one that does a full erase?

File: examples/lock-app/nrf5/BUILD.gn [hunk 3]
```diff
@@ -51,7 +51,100 @@ nrf5_sdk("sdk") {
...
+# Objcopy could also be wrapped in a platform-specific script to encapsulate
+# the target format.
+#
+# This is loosely inspired by Fuchsia's template("image_binary"). However,
+# Fuchsia uses a wrapper around GN toolchains to deal with the problem of
+# things like `objcopy` not being first-class tools in GN with composable
+# configuration, and that does not seem compatible with Pigweed's toolchain
+# utilities.
+
+template("flashable_executable") {
```
Comment (mspang) [inline examples/lock-app/nrf5/BUILD.gn:107]: isn't this specifically nrf5_executable?

File: examples/lock-app/nrf5/command_wrapper.py [hunk 4]
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env python
```
Comment (kpschoedel) [inline examples/lock-app/nrf5/command_wrapper.py:1]: Intend to replace this file by chromium's equivlanet, https://source.chromium.org/chromium/chromium/src/+/master:build/gn_run_binary.py

**Review by mspang — APPROVED** [review]
Now do efr32 which uses -O srec instead of -O ihex :)

**Comment by mspang** [comment]
Please move to a common file and also change the lighting app.

File: examples/lock-app/nrf5/make_flasher.py (hunk 1 above)
Comment (kpschoedel) [reply examples/lock-app/nrf5/make_flasher.py:37]: Will remove; was tracking my edits.

**Comment by kpschoedel** [comment]
> Now do efr32 which uses -O srec instead of -O ihex :)

Yes, that's exactly the problem with objcopy not being a config-able first-class tool.

File: examples/lock-app/nrf5/BUILD.gn (hunk 3 above)
Comment (kpschoedel) [reply examples/lock-app/nrf5/BUILD.gn:107]: There's nothing really nrf5-specific about it. Many platforms need the objcopy step (or equivalent) and all (?) need the flash-script-builder part. Or, are you suggesting an nrf5_executable layer to contain the platform-specific details like -Oihex?

File: examples/lock-app/nrf5/make_flasher.py (hunk 2 above)
Comment (kpschoedel) [reply examples/lock-app/nrf5/make_flasher.py]: Done

**Comment by mspang** [comment]
1) One step seems easier than 2-3. Could we generate one script per target, and accept arguments, e.g.

  `./out/debug/chip-nrf52840-lock-example.flash.py [--eraseall] [--skip-softdevice]`

and that one script erases (if requested), flashes the softdevice (maybe unless --skip-softdevice passed?), flashes the application, then resets the device?

2) Can we move the guts to a common file?

e.g. have the generated file be simply a function call something like

```
SCRIPTS_DIR = ""
FIRMWARE_PATH = ""
SOFTDEVICE_PATH = ""

import sys
sys.path.append(SCRIPTS_DIR)
import nrf5_firmware_utils

if __name__ == '__main__':
  nrf5_firmware_utils.flash(FIRMWARE_PATH, SOFTDEVICE_PATH)
```

(maybe we should even copy the softdevice .hex file into the output dir to make SOFTDEVICE_PATH a saner path)

3) can you run` gn_build.sh` and fix the errors?

**Comment by turon** [comment]
Doesn't this essentially boil down to two command lines?

```
arm-none-eabi-objcopy -O ihex $(<) $(APP_FILE_HEX)
nrfjprog -f NRF52 --sectorerase  --reset --program $(APP_FILE_HEX)
```

Would it be possible to simplify?

**Comment by rwalker-apple** [comment]
> Strawman for one method of building flashable binaries and a consistent
> method of flashing them, for parity with some existing `make flash`
> commands and ultimately for automated testing.
> 
> This is temporarily wholly inside examples/lock-app/nrf5 in order to
> experiment with minimal disruption but is intended to become generic.
> 
> issue #1520 GN build parity - flashing
> 
> #### Problem
> #### Summary of Changes
> <what's in this PR>
> 
> fixes #<issue number, or numbers. if no issue, please create one>

Hi @kpschoedel.  Thank you for this contribution!  Would you mind terribly filling out the PR template?

#### Problem
_<Please describe the problem you're addressing in a way that makes it easy for reviewers to identify how your changes  might impact the project.>_

#### Summary of Changes
_<Please describe what you had to touch and how in order to address the problem(s) above>_

**Review by andy31415 — APPROVED** [review]

**Comment by kpschoedel** [comment]
> 1. One step seems easier than 2-3. Could we generate one script per executable, and accept arguments, e.g.

Sure. I had it separate because I was aiming for the .flash.py script having nothing platform-dependent.

> 1. Can we move the guts to a common file?

Sure, it will just take me a moment to learn python packaging.

> (maybe we should even copy the softdevice .hex file into the output dir to make SOFTDEVICE_PATH a saner path)

That's more difficult because the softdevice.hex file doesn't have a fixed name (in Make it's `$(wildcard $(SOFTDEVICE_IMAGE_DIR)/s140_nrf52_*_softdevice.hex)` so GN can't have a dependency on it. I think in GN this requires another python script action, presumably taking the $(SOFTDEVICE_IMAGE_DIR) as the input dependency — do you know a better way?

**Comment by kpschoedel** [comment]
> Doesn't this essentially boil down to two command lines?
> 
> ```
> arm-none-eabi-objcopy -O ihex $(<) $(APP_FILE_HEX)
> nrfjprog -f NRF52 --sectorerase  --reset --program $(APP_FILE_HEX)
> ```
> 
> Would it be possible to simplify?

Unfortunately, I don't think so. In Make it's a few lines but in GN only programs that are part of its built-in concept of a toolchain are easy to run. So the goal here is to hide the effort under //gn so that at least a program's BUILD.gn is simple.

**Comment by mspang** [comment]
> > Doesn't this essentially boil down to two command lines?
> > ```
> > arm-none-eabi-objcopy -O ihex $(<) $(APP_FILE_HEX)
> > nrfjprog -f NRF52 --sectorerase  --reset --program $(APP_FILE_HEX)
> > ```
> > 
> > 
> > Would it be possible to simplify?
> 
> Unfortunately, I don't think so. In Make it's a few lines but in GN only programs that are part of its built-in concept of a toolchain are easy to run. So the goal here is to hide the effort under //gn so that at least a program's BUILD.gn is simple.



> > 1. One step seems easier than 2-3. Could we generate one script per executable, and accept arguments, e.g.
> 
> Sure. I had it separate because I was aiming for the .flash.py script having nothing platform-dependent.
> 
> > 1. Can we move the guts to a common file?
> 
> Sure, it will just take me a moment to learn python packaging.
> 
> > (maybe we should even copy the softdevice .hex file into the output dir to make SOFTDEVICE_PATH a saner path)
> 
> That's more difficult because the softdevice.hex file doesn't have a fixed name (in Make it's `$(wildcard $(SOFTDEVICE_IMAGE_DIR)/s140_nrf52_*_softdevice.hex)` so GN can't have a dependency on it. I think in GN this requires another python script action, presumably taking the $(SOFTDEVICE_IMAGE_DIR) as the input dependency — do you know a better way?

It has a fixed name, s140_nrf52_7.0.1_softdevice.hex

Could be an argument in case someone wants to override (the right arg might be "SDK version")

**Comment by kpschoedel** [comment]
> It has a fixed name, s140_nrf52_7.0.1_softdevice.hex
> 
> Could be an argument in case someone wants to override (the right arg might be "SDK version")

OK, I'll do that and strip out the Makefile-inspired globbing.

**Comment by mspang** [comment]
> > > Doesn't this essentially boil down to two command lines?
> > > ```
> > > arm-none-eabi-objcopy -O ihex $(<) $(APP_FILE_HEX)
> > > nrfjprog -f NRF52 --sectorerase  --reset --program $(APP_FILE_HEX)
> > > ```
> > > 

Let's compare apples to apples here. Are these the same problem?

 - run the programmer command to flash a single executable in the context of a makefile that builds a single application
 - generate a flashing helper for every embedded executable in the build, which includes multiple applications across a heterogeneous set of MCUs?

If we want 2 line approach, then we can certainly check in a 2 line script and call it done. But I think we'll quickly get tired of copy pasting 2 line scripts around (and in reality, this is already more complicated than above - there's the softdevice, etc, and maybe some error handling is warranted e.g. device isn't plugged in, do you need to go install some udev rules?).

> > > 
> > > Would it be possible to simplify?
> > 
> > 
> > Unfortunately, I don't think so. In Make it's a few lines but in GN only programs that are part of its built-in concept of a toolchain are easy to run.
> 
> > > 1. One step seems easier than 2-3. Could we generate one script per executable, and accept arguments, e.g.
> > 
> > 
> > Sure. I had it separate because I was aiming for the .flash.py script having nothing platform-dependent.
> > > 1. Can we move the guts to a common file?
> > 
> > 
> > Sure, it will just take me a moment to learn python packaging.
> > > (maybe we should even copy the softdevice .hex file into the output dir to make SOFTDEVICE_PATH a saner path)
> > 
> > 
> > That's more difficult because the softdevice.hex file doesn't have a fixed name (in Make it's `$(wildcard $(SOFTDEVICE_IMAGE_DIR)/s140_nrf52_*_softdevice.hex)` so GN can't have a dependency on it. I think in GN this requires another python script action, presumably taking the $(SOFTDEVICE_IMAGE_DIR) as the input dependency — do you know a better way?
> 
> It has a fixed name, s140_nrf52_7.0.1_softdevice.hex
> 
> Could be an argument in case someone wants to override (the right arg might be "SDK version")

**Review by rwalker-apple — APPROVED** [review]

**Review by mspang — APPROVED** [review]

**Comment by mspang** [comment]
@saurabhst @BroderickCarlin @hawk248 @jelderton ?

**Review by saurabhst — APPROVED** [review]

**Pull request merged**